lib/rules/no-this-in-before-route-enter.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * @fileoverview Don't use this in a beforeRouteEnter method
+ * @author Przemyslaw Jan Beigert
+ */
+'use strict'
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const utils = require('../utils')
+
+;('use strict')
+/**
+ * @param {FunctionExpression | BlockStatement | Property | ThisExpression} obj
+ * @returns {boolean}
+ */
+function deepFindThisExpression(obj) {
+  if (typeof obj !== 'object' || !obj) {
+    return false
+  }
+  if (obj.type === 'ThisExpression') {
+    return true
+  }
+
+  /** @param {typeof obj} key */
+  return Object.entries(obj).some(([key, value]) => {
+    if (key === 'parent') {
+      return false
+    }
+    if (Array.isArray(value)) {
+      return value.some((item) => deepFindThisExpression(item))
+    }
+    if (typeof value === 'object') {
+      return deepFindThisExpression(value)
+    }
+
+    return false
+  })
+}
+
+/**
+ * @param {Property | SpreadElement} property
+ * @returns {property is Property}
+ */
+function isPropertyBeforeRouteMethod(property) {
+  if (property.type !== 'Property') {
+    return false
+  }
+
+  return (
+    property.key.type === 'Identifier' &&
+    property.key.name === 'beforeRouteEnter'
+  )
+}
+
+const errorMessage =
+  'beforeRouteEnter does NOT have access to `this` component instance. https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards'
+
+module.exports = {
+  errorMessage,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow this usage in a beforeRouteEnter method',
+      categories: null,
+      url: 'https://eslint.vuejs.org/rules/no-this-in-before-route-enter.html'
+    },
+    schema: []
+  },
+
+  /** @param {RuleContext} context */
+  create(context) {
+    // ----------------------------------------------------------------------
+    // Public
+    // ----------------------------------------------------------------------
+    return utils.executeOnVue(context, (obj) => {
+      const beforeRouteProperty = obj.properties.find(
+        isPropertyBeforeRouteMethod
+      )
+      if (!beforeRouteProperty) {
+        return
+      }
+      if (beforeRouteProperty.value.type !== 'FunctionExpression') {
+        return
+      }
+      if (deepFindThisExpression(beforeRouteProperty.value.body)) {
+        context.report({
+          node: beforeRouteProperty,
+          message: errorMessage
+        })
+      }
+    })
+  }
+}
```

Comment:
Use `':function'` and `':function:exit'` like the `no-deprecated-props-default-this` rule to track `this`.

```suggestion
'use strict'

// ------------------------------------------------------------------------------
// Requirements
// ------------------------------------------------------------------------------

const utils = require('../utils')

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: 'problem',
    docs: {
      description: 'disallow this usage in a beforeRouteEnter method',
      categories: null,
    },
    fixable: null,
    schema: [],
    messages: {
      disallow:
        "'beforeRouteEnter' does NOT have access to `this` component instance. https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards."
    }
  },
  /** @param {RuleContext} context */
  create(context) {
    /**
     * @typedef {object} ScopeStack
     * @property {ScopeStack | null} upper
     * @property {FunctionExpression | FunctionDeclaration} node
     * @property {boolean} beforeRouteEnter
     */
    /** @type {Set<FunctionExpression>} */
    const beforeRouteEnterFunctions = new Set()
    /** @type {ScopeStack | null} */
    let scopeStack = null

    /**
     * @param {FunctionExpression | FunctionDeclaration | ArrowFunctionExpression} node
     */
    function onFunctionEnter(node) {
      if (node.type === 'ArrowFunctionExpression') {
        return
      }
      scopeStack = {
        upper: scopeStack,
        node,
        beforeRouteEnter: beforeRouteEnterFunctions.has(
          /** @type {never} */ (node)
        )
      }
    }

    /**
     * @param {FunctionExpression | FunctionDeclaration | ArrowFunctionExpression} node
     */
    function onFunctionExit(node) {
      if (scopeStack && scopeStack.node === node) {
        scopeStack = scopeStack.upper
      }
    }
    return utils.defineVueVisitor(context, {
      onVueObjectEnter(node) {
        const beforeRouteEnter = utils.findProperty(node, 'beforeRouteEnter')
        if (
          beforeRouteEnter &&
          beforeRouteEnter.value.type === 'FunctionExpression'
        ) {
          beforeRouteEnterFunctions.add(beforeRouteEnter.value)
        }
      },
      ':function': onFunctionEnter,
      ':function:exit': onFunctionExit,
      ThisExpression(node) {
        if (scopeStack && scopeStack.beforeRouteEnter) {
          context.report({
            node,
            messageId: 'disallow'
          })
        }
      }
    })
  }
}
```